kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/encryption/EnvelopeEncryptionFilterIT.java

```diff
@@ -248,6 +257,77 @@ void produceAndConsumeEncryptedAndPlainTopicsAtSameTime(KafkaCluster cluster, To
         }
     }
 
+    /**
+     * Test that ensures that the record offsets returned by the broker are faithfully relayed to the client.
+     * @param cluster underlying kafka cluster
+     * @param compactedTopic topic configured for compaction.
+     * @param directConsumer consumer connected directly to the underlying kafka cluster
+     * @param testKmsFacade kms facade
+     * @throws Exception exception
+     */
+    @TestTemplate
+    @SuppressWarnings("java:S2925")
+    void offsetFidelity(@BrokerConfig(name = "log.cleaner.backoff.ms", value = "500") @BrokerConfig(name = "log.retention.check.interval.ms", value = "500") KafkaCluster cluster,
+                        @TopicConfig(name = "segment.ms", value = "250") @TopicConfig(name = "cleanup.policy", value = "compact") Topic compactedTopic,
+                        Consumer<String, String> directConsumer,
+                        TestKmsFacade<?, ?, ?> testKmsFacade)
+            throws Exception {
+        var testKekManager = testKmsFacade.getTestKekManager();
+        testKekManager.generateKek(compactedTopic.name());
+
+        var builder = proxy(cluster);
+
+        builder.addToFilters(buildEncryptionFilterDefinition(testKmsFacade));
+
+        try (var tester = kroxyliciousTester(builder);
+                var proxyProducer = tester.producer();
+                var proxyConsumer = tester.consumer()) {
+
+            proxyProducer.send(new ProducerRecord<>(compactedTopic.name(), "a", "a1"));
+            // Send two messages for key "b", the first will be eligible for compaction
+            proxyProducer.send(new ProducerRecord<>(compactedTopic.name(), "b", "b1"));
+            proxyProducer.send(new ProducerRecord<>(compactedTopic.name(), "b", "b2")).get(5, TimeUnit.SECONDS);
+
+            // Sleep for > segment.ms to that the next segment will begin and first will become eligible for compaction
+            Thread.sleep(250);
```

Comment:
This rests on the assumption that the log cleaner thread will definitely run to completion at least once in 250ms. That's definitely not guaranteed, so this is likely to be flaky. We can make it _less_ flaky is we use something like 2×segment.ms. For higher self-documentation value we could factor out an `awaitCompaction` method, even if is just calls sleep right now.

Reply:
The intent of the code is to allow sufficient time for the broker to be obligated to start a new segment.  Is the log cleaner really involved in this in this part?  I looked at `kafka.log.UnifiedLog#maybeRoll` which appeared to me to be called on the append path.  I noticed the method is called by a thread 'data-plane-kafka-request-handler-x' which appears to be serving requests rather than have a cleaning role.   _I don't know the Kafka code well so I could very well be misunderstanding the broker under the debugger or making spurious inferences._ 

I think where I do need to wait for the log cleaner to act is the next part.  Here I've used Awaitility to consume repeatedly until the records appear in the 'right' compacted shape.

I think therefore my Thread.sleep(segment.ms) is okay.  Sufficient time will have passed so the Broker is forced to start a new segment before it can write `c`. 

My application of `log.retention.check.interval.ms` is superfluous.  Log retention has no part to play in this text.

Reply:
OK, yes, I think you're right